soufi/finders/yum.py

```diff
@@ -0,0 +1,192 @@
+            if not repo_url.endswith('/'):
+                repo_url += '/'
+            repo = self._get_repo(repo_url)
+            if repo is None:
+                continue
+            for package in repo.findall(name):
+                # If the package version in the repomd is our version,
+                # it's easy.  Note that we want to match epoch-full and
+                # epoch-less version formats.
+                if version in (package.evr, package.vr):
+                    return repo.baseurl + package.location
+                # Otherwise let's make it weird
+                packages.append(package)
+
+            # If we've made it here, things have gotten weird.  Replace the
+            # version+release info in all unique package locations with our
+            # version and see if they exist.  This should find superseded
+            # packages that are present in the repo, but not in the repomd.
+            locs = set(p.location.replace(p.vr, version) for p in packages)
+            for loc in locs:
+                if self._test_url(repo.baseurl + loc):
+                    return repo.baseurl + loc
+        return None
+
+    def _walk_binary_repos(self, name):
+        packages = []
+        for repo_url in self.binary_repos:
+            repo = self._get_repo(repo_url)
+            if repo is None:
+                continue
+            for package in repo.findall(name):
+                # If we have a binary package matching our version, but with
+                # a different name, return the NVR fields
+                if self.version in (package.evr, package.vr):
+                    return self._nevra_or_none(package)
+                # Otherwise let's make it weird
+                packages.append(package)
+
+        # If we've made it here, things have gotten weird; this should be
+        # the case of a source RPM that produces binary RPMs with different
+        # names *and* versions (the lvm2 package from Red Hat is a good example
+        # of this).  In this case, we won't be able to make heads or tails of
+        # the response, unless (and only unless) it contains a single package.
+        try:
+            [package] = packages
+        except ValueError:
+            return None, None
+        return self._nevra_or_none(package)
+
+    def _nevra_or_none(self, package):
+        if package.sourcerpm == '':
+            # It's here, but has no sources defined!  Bummer...
+            return None, None
+        nevra = self._get_nevra(package.sourcerpm)
+        return nevra['name'], f"{nevra['ver']}-{nevra['rel']}"
+
+    # Cache repo downloads as they are slow and network-bound.
+    @classmethod
+    @functools.lru_cache(maxsize=128)
+    def _get_repo(cls, url):
+        try:
+            return repomd.load(url)
+        except urllib.error.HTTPError:
+            return None
+
+    @staticmethod
+    def _get_nevra(filename):
+        """Split out the NEVRA fields from an RPM filename."""
+
```

Comment:
Surely there's a library that does this? There's a bunch of stuff in the yum python module that makes me think we're missing out a bit.

Reply:
There is, but I recall it having some issues, so I stole the useful bits and reimplemented it how I liked it

Reply:
Ok that needs justifying in a comment please. Otherwise we'll come back here in a year and think "why didn't this idiot just use X library" ...